vpr/src/pack/post_routing_pb_pin_fixup.cpp

```diff
@@ -0,0 +1,637 @@
+                pin_sides.push_back(pinloc_side);
+            }
+        }
+        /* We should have at least one side now after merging */
+        VTR_ASSERT(!pin_sides.empty());
+
+        ClusterNetId routing_net_id = ClusterNetId::INVALID();
+        std::vector<RRNodeId> visited_rr_nodes;
+        short valid_routing_net_cnt = 0;
+        for (const e_side& pin_side : pin_sides) {
+            /* Find the net mapped to this pin in routing results */
+            const int& rr_node = get_rr_node_index(device_ctx.rr_node_indices,
+                                                   grid_coord.x(), grid_coord.y(),
+                                                   rr_node_type, physical_pin, pin_side);
+
+            /* Bypass invalid nodes, after that we must have a valid rr_node id */
+            if (OPEN == rr_node) {
+                continue;
+            }
+            VTR_ASSERT((size_t)rr_node < device_ctx.rr_nodes.size());
+
+            /* If the node has been visited on the other side, we just skip it */
+            if (visited_rr_nodes.end() != std::find(visited_rr_nodes.begin(), visited_rr_nodes.end(), RRNodeId(rr_node))) {
+                continue;
+            }
+
+            /* Get the cluster net id which has been mapped to this net
+             * In general, there is only one valid rr_node among all the sides.
+             * However, we have an exception in the Stratix-IV arch modeling,
+             * where a pb_pin may exist in two different sides but
+             * router will only map to 1 rr_node 
+             * Therefore, it is better to compare the routing nets
+             * for all the sides and pick
+             * - The unique valid net id (others should be all invalid)
+             *   assume that this pin is used by router
+             * - A invalid net id (others should be all invalid as well)
+             *   assume that this pin is not used by router
+             */
+            if (rr_node_nets[RRNodeId(rr_node)]) {
+                if (routing_net_id) {
+                    if (routing_net_id != rr_node_nets[RRNodeId(rr_node)]) {
+                        VTR_LOG_ERROR("Pin '%s' is mapped to two nets: '%s' and '%s'\n",
+                                      pb_graph_pin->to_string().c_str(),
+                                      clustering_ctx.clb_nlist.net_name(routing_net_id).c_str(),
+                                      clustering_ctx.clb_nlist.net_name(rr_node_nets[RRNodeId(rr_node)]).c_str());
+                    }
+                    VTR_ASSERT(routing_net_id == rr_node_nets[RRNodeId(rr_node)]);
+                }
+                routing_net_id = rr_node_nets[RRNodeId(rr_node)];
+                valid_routing_net_cnt++;
+                visited_rr_nodes.push_back(RRNodeId(rr_node));
+            }
+        }
+        if (!((0 == valid_routing_net_cnt) || (1 == valid_routing_net_cnt)))
```

Comment:
`if (valid_routing_net_cnt) < 0 || valid_routing_net_cnt >= 2)` is easier to understand.